src/Nethermind/Nethermind.Consensus/Producers/TxPoolTxSource.cs

```diff
@@ -14,161 +15,291 @@
-                }
+            int checkedTransactions = selectedTxs.Count;
+
+            long blockNumber = parent.Number + 1;
+            // Obtain a transaction comparer based on the block context (e.g., gas price, fees)
+            IComparer<Transaction> comparer = GetComparer(parent, new BlockPreparationContext(baseFee, blockNumber))
+                .ThenBy(ByHashTxComparer.Instance); // in order to sort properly and not lose transactions we need to differentiate on their identity which provided comparer might not be doing
 
-                bool success = _txFilterPipeline.Execute(tx, parent);
-                if (!success) continue;
+            // Sort the selected transactions using the comparer
+            selectedTxs.AsSpan().Sort(comparer);
 
-                foreach (Transaction blobTx in PickBlobTxsBetterThanCurrentTx(selectedBlobTxs, tx, comparer))
-                {
-                    yield return blobTx;
-                }
+            // Sort transactions from the same sender by their nonce in ascending order
+            // Otherwise txs will be invalidated by not being in nonce order
+            SortMultiSendersByNonce(selectedTxs, parent, spec, token);
+            // Remove any transactions that have been invalidated or are no longer valid after sorting
+            RemoveBlankedTransactions(selectedTxs);
 
-                if (_logger.IsTrace) _logger.Trace($"Selected {tx.ToShortString()} to be potentially included in block.");
+            // Re-sort transactions to maintain the correct order, ensuring transactions from the same sender remain in nonce order
+            // As sorting by nonce may have put txs with lower priorty earlier
+            selectedTxs.AsSpan().Sort((tx0, tx1) =>
+                tx0.SenderAddress == tx1.SenderAddress ?
+                0 :  // If transactions are from the same sender, keep their current order
+                comparer.Compare(tx0, tx1)); // Otherwise, sort based on the main comparer
 
-                selectedTransactions++;
-                yield return tx;
-            }
+            // Remove blob transactions that cannot be included in the block (e.g. due to block limit constraints)
+            RemoveBlobTransactions(selectedTxs.AsSpan(), parent, spec);
+            // Remove any additional invalidated transactions after blob transaction removal
+            RemoveBlankedTransactions(selectedTxs);
 
-            if (selectedBlobTxs.Count > 0)
+            // Convert the list of selected transactions to an array for inclusion in the block
+            Transaction[] potentialTxs = selectedTxs.ToArray();
+
+            if (_logger.IsDebug) _logger.Debug($"Potentially selected {potentialTxs.Length} out of {checkedTransactions} pending transactions checked.");
+
+            // Return the selected transactions unless the operation was cancelled
+            return token.IsCancellationRequested ? Array.Empty<Transaction>() : potentialTxs;
+        }
```

Comment:
Really dislike the solution:
1. Sort things multiple times
2. Allocating memory while previous solution was lazy.
3. Prioritizing GasPrice sorting over Nonce sorting first.

`GetOrderedTransactions` where close to optimal, lazy and more readable solution and I would prefer to keep it.